src/primitives/transaction.cpp

```diff
@@ -96,18 +96,9 @@ class SproutProofVerifier : public boost::static_visitor<bool>
 
     bool operator()(const libzcash::PHGRProof& proof) const
     {
-        return params.verify(
-            proof,
-            verifier,
-            joinSplitPubKey,
-            jsdesc.randomSeed,
-            jsdesc.macs,
-            jsdesc.nullifiers,
-            jsdesc.commitments,
-            jsdesc.vpub_old,
-            jsdesc.vpub_new,
-            jsdesc.anchor
-        );
+        // We checkpoint after Sapling activation, so we can skip verification
+        // for all Sprout proofs.
```

Comment:
Is this strictly true? I thought that the checkpoint mechanism didn't enforce that the last checkpoint is actually reached. What happens in this case if a node that is syncing sees a block with an invalid Sprout proof (that is not on the main chain, but it doesn't know that yet)?

(This is a blocking question.)

Reply:
Also relevant: https://github.com/zcash/zcash/issues/3996 (Prevent creation of shielded transactions when chain is not synced up).

Reply:
I've looked more closely into the current node behaviour.

The checkpoint hashes in the chain parameters are only directly used in [`Checkpoints::GetLastCheckpoint`](https://github.com/zcash/zcash/blob/a722aa69a2b5e9c94731f6c5d268b298f247aace/src/checkpoints.cpp#L67-L79). It is called in `main.cpp` in two places: in `ConnectBlock` (to disable script verification), and in `ContextualCheckBlockHeader` to enforce no pre-checkpoint forking:

https://github.com/zcash/zcash/blob/a722aa69a2b5e9c94731f6c5d268b298f247aace/src/main.cpp#L3900-L3906

This relies on the fact that `GetLastCheckpoint` returns the most recent checkpoint _that we have received a header for_. Thus if it returns non-null, it implies we have already called `ContextualCheckBlockHeader` for all block headers up to and including that checkpoint, meaning that this is a call with some non-main-chain header.

Block headers are fetched by the node sequentially: [peers are required to provide contiguous headers](https://github.com/zcash/zcash/blob/a722aa69a2b5e9c94731f6c5d268b298f247aace/src/main.cpp#L5966-L5969), and [`AcceptBlockHeader` returns an error if a header's parent is not known locally](https://github.com/zcash/zcash/blob/a722aa69a2b5e9c94731f6c5d268b298f247aace/src/main.cpp#L3998-L4007).

The `fCheckpointsEnabled` flag [defaults to `true`](https://github.com/zcash/zcash/blob/a722aa69a2b5e9c94731f6c5d268b298f247aace/src/init.cpp#L977), but can be configured at start with `-checkpoints`; it is [documented only in terms of disabling script verification](https://github.com/zcash/zcash/blob/a722aa69a2b5e9c94731f6c5d268b298f247aace/src/init.cpp#L451), but this makes it likely that most users (wanting speed) are running with the default.

Besides the above, there is also the `nMinimumChainWork` consensus parameter. A node [will not leave initial block download mode until this chain work is reached](https://github.com/zcash/zcash/blob/a722aa69a2b5e9c94731f6c5d268b298f247aace/src/main.cpp#L1853-L1854). A node [will not advertise blocks to peers while it is in initial block download mode](https://github.com/zcash/zcash/blob/a722aa69a2b5e9c94731f6c5d268b298f247aace/src/main.cpp#L3434-L3450).

Currently in `zcashd` we have two checkpoint hashes after Sapling activation, and `nMinimumChainWork` was last updated in the 2.0.7 release, with a value from [block 588210](https://github.com/zcash/zcash/pull/4108#issuecomment-522759484) (after Sapling activation).

---

Given all of the above, I think the current state of play is that before a node has received all main chain block headers prior to Sapling activation, it can be fed a fake non-main chain with invalid old Sprout proofs (for any block height region between the highest main chain checkpoint header the node knows, and Sapling activation), and it would accept the fake chain as valid. However:
- The node would be non-functional (and would not broadcast the fake chain to other peers) until the fake chain contained as much work as the main chain. Sapling activation was almost a year ago, so this would likely be costly (but I have not figured out what this cost would be).
- The node would happily reorg away from the invalid chain to the main chain as it received main chain blocks (unless the fake chain had more chain work). Thus an adversary that does not have eclipse power would only be able to increase the bandwidth and storage costs of the node, which it can already do in far cheaper ways (that don't require faking an entire chain).

I think this therefore reduces to a 50%+1 attack.

Reply:
Re: #3996, I have not addressed it completely in this PR, but after this PR shielded transactions now cannot be created at all before Sapling activates (this is made explicit while removing the `makeGrothProof ` parameter).

Reply:
Conceptually there seems to be a difference between script validation and proof validation: an attacker who can subvert script validation can steal money but not violate overall balance. An attacker who can subvert proof validation can violate overall balance. On the other hand, I'm not sure this matters if the only available attacks are during IBD *and* require more work than a 51% attack.

I would really like to fix #3996, but I don't consider it a blocker for merging this PR.

Reply:
The suggestion in https://github.com/zcash/zcash/pull/4060#discussion_r326182464 (aborting if we're on a chain with at least the minimum work that doesn't activate the last NU at the correct block hash) would address my concern here.

Reply:
Oh, I hadn't realized that we *already* disable proof verification for blocks before a checkpoint, exactly when we disable script verification (https://github.com/zcash/zcash/blob/master/src/main.cpp#L2599). I think that aborting if we're clearly on the wrong chain is still the right approach, though.